pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
-    "mocha": "<current-version>"
+    "@loopback/testlab": "^<current-version>",
+    "mocha": "^<current-version>"
   },
   "scripts": {
     "test": "mocha"
   }
-  ...
+  // ...
+}
 ```
 
+## Unit testing
+
+Unit tests are considered as "white-box". They use "inside-out" approach, where the test knows all about the internals and controls all the variables of the system under test. Individual units are tested in isolation, their dependencies are replaced with [Test doubles](https://en.wikipedia.org/wiki/Test_double).
+
+### Use test doubles
+
+Test doubles are functions or objects that look and behave like the real variants used in production, but are actually simplified versions giving the test more control of the behavior. For example, reproducing the situation where reading from a file failed because of a hard-drive error is pretty much impossible, unless we are using a test double that's simulating file-system API and giving us control of how what each call returns.
+
+[Sinon.JS](http://sinonjs.org/) has become the de-facto standard for Test doubles in Node.js and JavaScript/TypeScript in general. We recommend our users to leverage this library, our `@loopback/testlab` package comes with Sinon preconfigured with TypeScript type definitions and integrated with Should.js assertions.
+
+There are three kinds of Test doubles provided by Sinon.JS:
+
+- [Test spies](http://sinonjs.org/releases/v4.0.1/spies/) are functions that record arguments, return value, the value of `this` and exception thrown (if any) for all its calls. There are two types of spies: Some are anonymous functions, while others wrap methods that already exist in the system under test.
+
+- [Test stubs](http://sinonjs.org/releases/v4.0.1/stubs/) are functions (spies) with pre-programmed behavior. As spies, stubs can be either anonymous, or wrap existing functions. When wrapping an existing function with a stub, the original function is not called.
+
+- [Test mocks](http://sinonjs.org/releases/v4.0.1/mocks/)  (and mock expectations) are fake methods (like spies) with pre-programmed behavior (like stubs) as well as pre-programmed expectations. A mock will fail your test if it is not used as expected.
+
+{% include note.html content="We recommend against using test mocks. With test mocks, the expectations must be defined before the tested scenario is executed, which breaks the recommended test layout 'arrange-act-assert' (or  'given-when-then') and produces code that's difficult to comprehend.
+" %}
+
+#### Create a stub Repository
+
+When writing an application accessing data in a database, it's a best practice to use [Repositories](./Repositories.html) to encapsulate all data-access/persistence-related code and let other parts of the application (typically [Controllers](./Controllers.html)) to depend on these Repositories for data access. In order to test Repository dependants (e.g. Controllers) in isolation, we need to provide a test double, usually as a test stub.
```

Comment:
Break up into multiple lines.

Reply:
`let other parts of the application (typically [Controllers](./Controllers.html)) to depend on these Repositories...`
should be
`let other parts of the application (typically [Controllers](./Controllers.html)) depend on these Repositories...`